src/test/kotlin/org/rust/lang/core/type/RsExpressionTypeInferenceTest.kt

```diff
@@ -476,4 +476,58 @@ class RsExpressionTypeInferenceTest : RsTypificationTestBase() {
             foo(|t| { t.bar(); })
         }           //^ &T
     """)
+
+    fun `test associated types for impl`() = testExpr("""
+        trait A {
+            type Item;
+            fn foo(self) -> Self::Item;
+        }
+        struct S;
+        impl A for S {
+            type Item = S;
+            fn foo(self) -> Self::Item {S}
```

Comment:
Let's add spaces here: `-> Self::Item { S }`. Sorry that rust autoformatting does not work with langauge injection :) !